### PR TITLE
Improve error handling of experimental features in non-experimental mode

### PR DESCRIPTION
**Pull request by yongtang**
**- What I did**

This fix tries to address several issues raised in #28626 where run against a non-experimental daemon may not generate correct error message:
1. Incorrect flags were not checked against the supported features:
   ```
   $ docker stack --nonsense
   unknown flag: --nonsense
   ```
2. Subcommands were not checked against the supported features:
   ```
   $ docker stack ls
   Error response from daemon: This node is not a swarm manager...
   ```

**- How I did it**

This fix address the above mentioned issues by:
1. Add a pre-check for FlagErrorFunc
2. Recursively check if a feature is supported for cmd and its parents.

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

misfondoshd com_hermoso-gatito-recostado_1366x768

This fix related to #28626. There is on remaining issue in #28626 where a feature was not checked to see if it is supported:
```
$ docker deploy
"docker deploy" requires exactly 1 argument(s).
```

This one is handled in `Args`. Will investigate further to see if it is possible to fix this specific issue as well.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by yongtang** [comment]
cc @albers @thaJeztah @vieux @mlaventure

**Comment by yongtang** [comment]
There seems to be some issues with the interaction between daemon and client. Let me close this PR for now.

Will reopen if I could fix the issue.

**Comment by yongtang** [comment]
@albers @thaJeztah @vieux @mlaventure The issue has been fixed. Please take a look.

**Comment by albers** [comment]
Thanks @yongtang. I'm sorry, but there are still invocations that produce an error message where I would expect them not to do so:
- `docker stack hugo`
- `docker checkpoint create`
- `docker checkpoint hugo`
- `docker deploy`

**Comment by yongtang** [comment]
Thanks @albers. The PR has been updated. Please take a look and let me know if there are any issues.

**Comment by albers** [comment]
Seems `docker stack` is no longer experimental, so the command now behaves like any non-experimental command. => Everything OK with this command in this PR.
With the other commands: I found no issues, everything works as expected. Thanks, @yongtang!
LGTM

File: cmd/docker/docker.go [hunk 1]
```diff
@@ -82,9 +99,53 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
...
+	})
+
 	return cmd
 }
 
+// visitAll will traverse all commands from the root.
+// This is different from the VisitAll of cobra.Command where only parents
+// are checked.
+func visitAll(root *cobra.Command, fn func(*cobra.Command)) {
+	var traverse func(r *cobra.Command, fn func(*cobra.Command))
```
Comment (LK4D4) [inline cmd/docker/docker.go]: this looks extremely weird to me. Why you need closure at all?

**Review by LK4D4 — CHANGES_REQUESTED** [review]

**Comment by LK4D4** [comment]
@yongtang posted comment

File: cmd/docker/docker.go (hunk 1 above)
Comment (yongtang) [inline cmd/docker/docker.go]: Thanks @LK4D4 for the review. The PR has been updated and `visitAll` has been simplified. Please take a look and let me know if there are any issues.

**Comment by yongtang** [comment]
Ping @LK4D4 @thaJeztah please take a look.

File: cmd/docker/docker.go [hunk 2]
```diff
@@ -52,6 +52,23 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
 		},
 	}
 	cli.SetupRootCommand(cmd)
+	// When invoking `docker stack --nonsense`, we need to make sure FlagErrorFunc return appropriate
+	// output if the feature is not supported.
+	// As above cli.SetupRootCommand(cmd) have already setup the FlagErrorFunc, we will add a pre-check before the FlagErrorFunc
+	// is called.
+	flagErrorFunc := cmd.FlagErrorFunc()
+	cmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
+		if dockerCli.Client() == nil { // when using --help, PersistenPreRun is not called, so initialization is needed.
```
Comment (aaronlehmann) [inline cmd/docker/docker.go]: typo in comment

File: cmd/docker/docker.go [hunk 3]
```diff
@@ -52,6 +52,23 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
...
+			dockerCli.Initialize(opts)
+		}
+		if err := isSupported(cmd, dockerCli.Client().ClientVersion(), dockerCli.HasExperimental()); err != nil {
+			return err
+		}
+		return flagErrorFunc(cmd, err)
+	})
 
 	cmd.SetHelpFunc(func(ccmd *cobra.Command, args []string) {
 		if dockerCli.Client() == nil { // when using --help, PersistenPreRun is not called, so initialization is needed.
```
Comment (aaronlehmann) [inline cmd/docker/docker.go]: typo in comment

File: cmd/docker/docker.go [hunk 4]
```diff
@@ -82,9 +99,47 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
...
+		// if there is no tags for a command or any of its parent,
+		// there is no need to wrap the Args validation.
+		if !hasTags(ccmd) {
+			return
+		}
+
+		cmdArgs := ccmd.Args
+		if cmdArgs != nil {
+			ccmd.Args = func(cmd *cobra.Command, args []string) error {
+				if dockerCli.Client() == nil { // when using --help, PersistenPreRun is not called, so initialization is needed.
```
Comment (aaronlehmann) [inline cmd/docker/docker.go]: typo in comment

**Comment by yongtang** [comment]
Thanks @aaronlehmann for the review. The PR has been updated.

**Comment by thaJeztah** [comment]
Both Janky and Experimental timed out on the same test, not sure if it's related or something else is causing this?

https://jenkins.dockerproject.org/job/Docker-PRs/38016/console

```
02:17:51 --- FAIL: Test (3448.65s)
02:17:51 panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s [recovered]
02:17:51 	panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s
02:17:51 
```

https://jenkins.dockerproject.org/job/Docker-PRs-experimental/29423/console

```
02:14:06 PASS: docker_cli_swarm_test.go:831: DockerSwarmSuite.TestUnlockEngineAndUnlockedSwarm	1.569s
02:19:06 --- FAIL: Test (3463.01s)
02:19:06 panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s [recovered]
02:19:06 	panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s
02:19:06 
```

**Comment by thaJeztah** [comment]
Hm, seen that same failure on another PR this morning, so should not be related; wondering what's causing it though

**Comment by yongtang** [comment]
@thaJeztah The PR was rebased to fix the merging conflict. Now all tests passed. Please take a look.

**Review by AkihiroSuda — COMMENTED** [review]
nit: now `newDockerCommand` is hard to follow because it has 110 lines.

Can you please split the function?

**Comment by yongtang** [comment]
Thanks @AkihiroSuda. The PR has been updated.

File: cmd/docker/docker.go [hunk 5]
```diff
@@ -166,9 +234,12 @@ func hideUnsupportedFeatures(cmd *cobra.Command, clientVersion string, hasExperi
 }
 
 func isSupported(cmd *cobra.Command, clientVersion string, hasExperimental bool) error {
-	if !hasExperimental {
-		if _, ok := cmd.Tags["experimental"]; ok {
-			return errors.New("only supported with experimental daemon")
+	// We check recursively so that, e.g., `docker stack ls` will return the same output as `docker stack`
+	for curr := cmd; curr != nil; curr = curr.Parent() {
```
Comment (LK4D4) [inline cmd/docker/docker.go]: can't you run this loop only if `!hasExperimental`?

File: cmd/docker/docker.go [hunk 6]
```diff
@@ -72,17 +111,46 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
...
+		// there is no need to wrap the Args validation.
+		if !hasTags(ccmd) {
+			return
+		}
 
-	cmd.SetOutput(dockerCli.Out())
-	cmd.AddCommand(newDaemonCommand())
-	commands.AddCommands(cmd, dockerCli)
+		cmdArgs := ccmd.Args
+		if cmdArgs != nil {
```
Comment (LK4D4) [inline cmd/docker/docker.go]: I'm not sure what `cmdArgs == nil` means, but doesn't it make sense to wrap it anyway and just return nil error?

**Comment by LK4D4** [comment]
Left some minor comments.

File: cmd/docker/docker.go (hunk 5 above)
Comment (yongtang) [inline cmd/docker/docker.go]: Thanks @LK4D4. Done.

File: cmd/docker/docker.go (hunk 6 above)
Comment (yongtang) [inline cmd/docker/docker.go]: @LK4D4 I changed the implementation so that if ccmd.Args (which is the func to check for number of Args) is nil, it will return immediately and remain as nil func.
This will reduce unnecessary scope in the previous implementation.

**Comment by yongtang** [comment]
Thanks @LK4D4 for the review. The PR has been updated. Please take a look.

File: cmd/docker/docker.go [hunk 7]
```diff
@@ -53,8 +53,47 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
...
+	// As above cli.SetupRootCommand(cmd) have already setup the FlagErrorFunc, we will add a pre-check before the FlagErrorFunc
+	// is called.
+	flagErrorFunc := cmd.FlagErrorFunc()
+	cmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
+		if dockerCli.Client() == nil { // when using --help, PersistentPreRun is not called, so initialization is needed.
+			// flags must be the top-level command flags, not cmd.Flags()
+			opts.Common.SetDefaultOptions(flags)
+			dockerPreRun(opts)
+			dockerCli.Initialize(opts)
+		}
```
Comment (dnephin) [inline cmd/docker/docker.go]: This block is now in 3 places. Can this be broken out into a new function that initializes the client?

File: cmd/docker/docker.go [hunk 8]
```diff
@@ -72,17 +111,48 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
...
+		if !hasTags(ccmd) {
+			return
+		}
 
-	cmd.SetOutput(dockerCli.Out())
-	cmd.AddCommand(newDaemonCommand())
-	commands.AddCommands(cmd, dockerCli)
+		if ccmd.Args == nil {
+			return
+		}
```
Comment (dnephin) [inline cmd/docker/docker.go]: The indentation looks off in github, some missing tabs?

Comment (yongtang) [inline cmd/docker/docker.go]: @dnephin The indentation seems to be OK. The issue is that diff is not able to find the logic `diff` piece:
```udiff
                        return
@@ -72,17 +100,52 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
                        ccmd.Println(err)
                }
        })
+}
 
-       flags = cmd.Flags()
-       flags.BoolVarP(&opts.Version, "version", "v", false, "Print version information and quit")
-       flags.StringVar(&opts.ConfigDir, "config", cliconfig.Dir(), "Location of client config files")
-       opts.Common.InstallFlags(flags)
+func setValidateArgs(dockerCli *command.DockerCli, cmd *cobra.Command, flags *pflag.FlagSet, opts *cliflags.ClientOptions) {
+       // The Args is handled by ValidateArgs in cobra, which does not allows a pre-hook.
+       // As a result, here we replace the existing Args validation func to a wrapper,
+       // where the wrapper will check to see if the feature is supported or not.
+       // The Args validation error will only be returned if the feature is supported.
+       visitAll(cmd, func(ccmd *cobra.Command) {
+               // if there is no tags for a command or any of its parent,
+               // there is no need to wrap the Args validation.
+               if !hasTags(ccmd) {
+                       return
+               }
 
-       cmd.SetOutput(dockerCli.Out())
-       cmd.AddCommand(newDaemonCommand())
-       commands.AddCommands(cmd, dockerCli)
+               if ccmd.Args == nil {
+                       return
+               }
 
-       return cmd
+               cmdArgs := ccmd.Args
```

File: cmd/docker/docker.go (hunk 7 above)
Comment (yongtang) [inline cmd/docker/docker.go]: Thanks @dnephin. Done.

**Comment by yongtang** [comment]
@dnephin The PR has been updated with function consolidated. Please take a look.

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by LK4D4** [comment]
@yongtang needs rebase :/ sorry

**Comment by yongtang** [comment]
Thanks @LK4D4. The PR has been rebased. Please take a look.

**Review by LK4D4 — APPROVED** [review]
LGTM
